File: matlab/tests/test_windowed_premaet.m
%% test_windowed_premaet.m — pre-MAET windowedSimilarity / windowedEntropy
%
%  Validates the pre-MAET windowedSimilarity and windowedEntropy against
%  the equivalent inline pipeline (weightEvents / translateAttributes /
%  buildExpTens / entropyExpTens / cosSimExpTens) they replace. Mirror of
%  Python's tests/test_windowed_premaet.py.
%
%  Standalone-runnable; appends to `results` when called from test_mpt.m.

if ~exist('results', 'var')
    results = {};
    standalone = true;
    addpath(fileparts(mfilename('fullpath')));
    clear cleanupDefaults
    cleanupDefaults = mptTestIsolateDefaults(); %#ok<NASGU>
else
    standalone = false;
end

tol   = 1e-9;
SD    = 1.0;
WW    = 2.0 * sqrt(3.0) * SD;          % variance-matched rectangular support
SIGP  = 0.12;
SIGT  = 0.05;

% Deterministic carrier (no RNG dependence across languages): a rising
% pitch line and a near-uniform onset grid.
N      = 24;
pitch  = (48 + (0:N-1)) ;                       % 1 x N rising pitches
onset  = cumsum(0.45 + 0.1 * mod((0:N-1), 3));  % 1 x N onsets
pAttr  = {pitch, onset};
centres = linspace(onset(1), onset(end), 9);

% query (a small triad template on pitch x time)
query  = {[60, 64, 67], [0, 0.5, 1.0]};
muQ    = mean(query{2});
qExt   = max(query{2}) - min(query{2});

% ----- 1. entropy, drop the time axis (differential & renyi2) -----
methodsShapes = {'differential', 0.0; 'renyi2', 0.0; 'renyi2', 1.0};
for k = 1:size(methodsShapes, 1)
    method = methodsShapes{k, 1};
    shape  = methodsShapes{k, 2};
    ref = zeros(1, numel(centres));
    for i = 1:numel(centres)
        if shape == 1.0
            [pw, ww] = weightEvents(pAttr, [], 2, 1, centres(i), shape, ...
                'width', WW, 'dropInputAttr', true);
        else
            [pw, ww] = weightEvents(pAttr, [], 2, 1, centres(i), shape, ...
                'sd', SD, 'dropInputAttr', true);
        end
        dens = buildExpTens(pw, ww, SIGP, 1, false, false, 0.0, 'verbose', false);
        ref(i) = entropyExpTens(dens, 'method', method, 'verbose', false);
    end
    got = windowedEntropy(pAttr, [], [SIGP, SIGT], [1, 1], [false, false], ...
        [false, false], [0.0, 0.0], centres, ...
        'contextWindow', {shape, WW}, 'method', method, ...
        'windowAttr', 2, 'dropWindowAttr', true, 'verbose', false);
    ok = max(abs(got(:) - ref(:))) < tol;
    results(end+1, :) = {sprintf('windowedEntropy drop window axis %s shape=%g', method, shape), ok}; %#ok<SAGROW>
end

% ----- 2. entropy, retain the time axis (joint pitch-time renyi2) --------
ref = zeros(1, numel(centres));
for i = 1:numel(centres)
    [pw, ww] = weightEvents(pAttr, [], 2, 1, centres(i), 1.0, ...
        'width', WW, 'dropInputAttr', false);
    dens = buildExpTens(pw, ww, [SIGP, SIGT], [1, 1], [false, false], ...
        [false, false], [0.0, 0.0], 'verbose', false);
    ref(i) = entropyExpTens(dens, 'method', 'renyi2', 'verbose', false);
end
got = windowedEntropy(pAttr, [], [SIGP, SIGT], [1, 1], [false, false], ...
    [false, false], [0.0, 0.0], centres, ...
    'contextWindow', {1.0, WW}, 'method', 'renyi2', 'windowAttr', 2, 'dropWindowAttr', false, 'verbose', false);
results(end+1, :) = {'windowedEntropy joint (retain axis)', ...
    max(abs(got(:) - ref(:))) < tol}; %#ok<SAGROW>

% ----- 3. similarity, locked template sweep (oneSidedDenom & cosine) -----
for nm = {'oneSidedDenom', 'cosine'}
    normalize = nm{1};
    ref = zeros(1, numel(centres));
    for i = 1:numel(centres)
        [pc, wc] = weightEvents(pAttr, [], 2, 1, centres(i), 1.0, ...
            'width', qExt, 'dropInputAttr', false);
        offs = {[], centres(i) - muQ};
        [pq, wq] = translateAttributes(query, [], offs);
        ref(i) = cosSimExpTens(pc, wc, pq, wq, [SIGP, SIGT], [1, 1], ...
            [false, false], [false, false], [0.0, 0.0], ...
            'normalize', normalize, 'verbose', false);
    end
    got = windowedSimilarity(pAttr, [], query, [], [SIGP, SIGT], [1, 1], ...
        [false, false], [false, false], [0.0, 0.0], centres, ...
        'normalize', normalize, 'windowAttr', 2, 'dropWindowAttr', false, 'verbose', false);
    ok = isequal(size(got), [1, numel(centres)]) && max(abs(got(:) - ref(:))) < tol;
    results(end+1, :) = {sprintf('windowedSimilarity locked %s', normalize), ok}; %#ok<SAGROW>
end

% ----- 4. similarity, decoupled 2-D correlogram (lag sweep) --------------
anchors = centres(2:8);
tau     = linspace(-1.0, 1.0, 9);
half    = 1.5;
ref = nan(numel(anchors), numel(tau));
for ia = 1:numel(anchors)
    a = anchors(ia);
    keep = abs(onset - a) <= half;
    pc = {pitch(keep), onset(keep)};
    for it = 1:numel(tau)
        offs = {[], (a - tau(it)) - muQ};
        [pq, wq] = translateAttributes(query, [], offs);
        ref(ia, it) = cosSimExpTens(pc, [], pq, wq, [SIGP, SIGT], [1, 1], ...
            [false, false], [false, false], [0.0, 0.0], ...
            'normalize', 'oneSidedDenom', 'verbose', false);
    end
end
qc2d = anchors(:) - tau(:).';        % nA x nTau absolute query centres
got = windowedSimilarity(pAttr, [], query, [], [SIGP, SIGT], [1, 1], ...
    [false, false], [false, false], [0.0, 0.0], anchors, ...
    'queryCentres', qc2d, 'contextWindow', {1.0, 2 * half}, ...
    'normalize', 'oneSidedDenom', 'windowAttr', 2, 'dropWindowAttr', false, 'verbose', false);
ok = isequal(size(got), size(ref)) && max(abs(got(:) - ref(:))) < tol;
results(end+1, :) = {'windowedSimilarity decoupled correlogram (2-D)', ok}; %#ok<SAGROW>

% ----- 5. generative sweep matches explicit centres ----------------------
explicit = onset(1):1.0:onset(end);
gExp = windowedSimilarity(pAttr, [], query, [], [SIGP, SIGT], [1, 1], ...
    [false, false], [false, false], [0.0, 0.0], explicit, ...
    'normalize', 'oneSidedDenom', 'windowAttr', 2, 'dropWindowAttr', false, 'verbose', false);
gGen = windowedSimilarity(pAttr, [], query, [], [SIGP, SIGT], [1, 1], ...
    [false, false], [false, false], [0.0, 0.0], [], ...
    'start', onset(1), 'stop', onset(end), 'step', 1.0, ...
    'normalize', 'oneSidedDenom', 'windowAttr', 2, 'dropWindowAttr', false, 'verbose', false);
ok = isequal(size(gGen), size(gExp)) && max(abs(gGen(:) - gExp(:))) < tol;
results(end+1, :) = {'windowedSimilarity generative == explicit', ok}; %#ok<SAGROW>

% ----- 6. error paths -----------------------------------------------------
gotR2 = windowedEntropy(pAttr, [], [SIGP, SIGT], [1, 2], [false, false], ...
    [false, false], [0.0, 0.0], centres, 'contextWindow', {1.0, WW}, ...
    'windowAttr', 2, 'dropWindowAttr', true, 'verbose', false);
okR2 = isequal(size(gotR2), [1, numel(centres)]) && all(isfinite(gotR2(:)));
results(end+1, :) = {'windowedEntropy drop r>=2 now allowed', okR2}; %#ok<SAGROW>

threwMarg = false;
try
    windowedEntropy(pAttr, [], [SIGP, SIGT], [1, 1], [false, false], ...
        [false, false], [0.0, 0.0], centres, 'contextWindow', {1.0, WW}, ...
        'windowAttr', 2, 'dropWindowAttr', false, 'marginalise', 1, ...
        'verbose', false);
catch
    threwMarg = true;
end
results(end+1, :) = {'windowedEntropy marginalise not implemented', threwMarg}; %#ok<SAGROW>

threwWidth = false;
try
    windowedEntropy(pAttr, [], [SIGP, SIGT], [1, 1], [false, false], ...
        [false, false], [0.0, 0.0], centres, 'contextWindow', {1.0, []}, ...
        'windowAttr', 2, 'dropWindowAttr', false, 'verbose', false);
catch
    threwWidth = true;
end
results(end+1, :) = {'windowedEntropy requires width', threwWidth}; %#ok<SAGROW>

threwBoth = false;
try
    windowedSimilarity(pAttr, [], query, [], [SIGP, SIGT], [1, 1], ...
        [false, false], [false, false], [0.0, 0.0], centres, ...
        'step', 1.0, 'windowAttr', 2, 'dropWindowAttr', false, 'verbose', false);
catch
    threwBoth = true;
end
results(end+1, :) = {'windowedSimilarity centres+step errors', threwBoth}; %#ok<SAGROW>

% ----- 7. single-axis == one-entry multi-axis sweep ---------------------
sgl = windowedSimilarity(pAttr, [], query, [], [SIGP, SIGT], [1, 1], ...
    [false, false], [false, false], [0.0, 0.0], centres, ...
    'normalize', 'oneSidedDenom', 'windowAttr', 2, 'dropWindowAttr', false, 'verbose', false);
mlt = windowedSimilarity(pAttr, [], query, [], [SIGP, SIGT], [1, 1], ...
    [false, false], [false, false], [0.0, 0.0], ...
    'sweep', {2, centres}, 'drop', {2, false}, ...
    'normalize', 'oneSidedDenom', 'verbose', false);
ok = isequal(size(sgl), size(mlt)) && max(abs(sgl(:) - mlt(:))) < tol;
results(end+1, :) = {'windowedSimilarity single == one-entry sweep', ok}; %#ok<SAGROW>

% ----- 8. multi-axis two-axis map (pitch swept+compared, time dropped) ---
patP = [60, 63, 60, 65]; patT = [0, 0.5, 1.5, 2.0];
PP = [patP, patP + 5]; TT = [patT, patT + 10];
[pb8, wb8, sb8] = bindEvents({PP, TT}, [], [4, 4], 'step', 1, 'relOuter', [false, false]);
[qb8, qw8, ~]   = bindEvents({patP, patT}, [], [4, 4], 'step', 1, 'relOuter', [false, false]);
R8 = windowedSimilarity(pb8, wb8, qb8, qw8, [SIGP, SIGT], [1, 1], [false, false], ...
    [false, false], [0.0, 0.0], 'sweep', {2, [1.0, 11.0]; 1, [62.0, 67.0]}, ...
    'drop', {2, true; 1, false}, 'specs', sb8, 'verbose', false);
ok = isequal(size(R8), [2, 2]) && R8(1,1) > 0.99 && R8(2,2) > 0.99 ...
     && R8(1,2) < 0.5 && R8(2,1) < 0.5;
results(end+1, :) = {'windowedSimilarity two-axis map', ok}; %#ok<SAGROW>

% ----- 9. locate is wired (centroid vs start peak at different centres) --
[qb9, qw9, qs9] = bindEvents({[60, 64], [0.0, 0.6]}, [], [1, 2], 'step', 1, 'relOuter', [false, false]);
cc9 = linspace(-0.4, 0.7, 12);
common9 = {'windowAttr', 2, 'dropWindowAttr', false, 'contextWindow', {1.0, 1.5}, ...
           'normalize', 'cosine', 'specs', qs9, 'verbose', false};
aC = windowedSimilarity(qb9, qw9, qb9, qw9, [SIGP, SIGT], [1, 2], [false, false], ...
    [false, false], [0.0, 0.0], cc9, 'locate', 'centroid', common9{:});
aS = windowedSimilarity(qb9, qw9, qb9, qw9, [SIGP, SIGT], [1, 2], [false, false], ...
    [false, false], [0.0, 0.0], cc9, 'locate', 'start', common9{:});
[~, iC] = max(aC); [~, iS] = max(aS);
ok = max(aC) > 0.9 && max(aS) > 0.9 && abs(cc9(iC) - cc9(iS)) > 0.2 && ~isequal(aC, aS);
results(end+1, :) = {'windowedSimilarity locate wired', ok}; %#ok<SAGROW>


if standalone
    nPass = sum([results{:, 2}]);
    fprintf('test_windowed_premaet: %d/%d passed\n', nPass, size(results, 1));
    for i = 1:size(results, 1)
        if ~results{i, 2}
            fprintf('  FAIL: %s\n', results{i, 1});
        end
    end
end

File: python/tests/test_windowed_premaet.py
"""Tests for the pre-MAET ``windowed_similarity`` and ``windowed_entropy``.

Each new-function result is checked against the equivalent inline pipeline
(``weight_events`` / ``translate_attributes`` / ``build_exp_tens`` /
``entropy_exp_tens`` / ``cos_sim_exp_tens``) it replaces, so the functions
are pinned to the hand-written composition rather than to remembered
numbers. Both argument surfaces are exercised: the single-axis form
(``window_attr`` / ``centres`` / ``drop_window_attr``, plus the decoupled
``query_centres`` correlogram) and the multi-axis form (``sweep`` / ``drop``).
"""

import numpy as np
import pytest

from mpt import (
    windowed_similarity, windowed_entropy,
    weight_events, translate_attributes, bind_events,
    unpack_pre_maet,
    build_exp_tens, entropy_exp_tens, cos_sim_exp_tens,
)

SD = 1.0
W_WIDTH = 2.0 * np.sqrt(3.0) * SD          # variance-matched rectangular support
SIG_P, SIG_T = 0.12, 0.05


@pytest.fixture
def triple():
    rng = np.random.default_rng(0)
    N = 40
    pitch = np.sort(rng.integers(48, 84, size=N).astype(float)).reshape(1, N)
    onset = np.cumsum(rng.uniform(0.4, 0.6, size=N)).reshape(1, N)
    return [pitch, onset], np.linspace(onset.min(), onset.max(), 9)


@pytest.fixture
def query():
    return [np.array([[60., 64., 67.]]), np.array([[0., 0.5, 1.0]])]


# --------------------------------------------------------------------------
# windowed_entropy
# --------------------------------------------------------------------------
@pytest.mark.parametrize("method,shape", [
    ("differential", 0.0), ("renyi2", 0.0), ("renyi2", 1.0),
])
def test_entropy_drop_window_axis(triple, method, shape):
    p_attr, centres = triple
    kw = {"width": W_WIDTH} if shape == 1.0 else {"sd": SD}
    ref = np.empty(len(centres))
    for i, c in enumerate(centres):
        pw, ww, _ = unpack_pre_maet(weight_events(p_attr, None, 1, 0, float(c), shape,
                                  is_per=False, period=0.0, drop_input_attr=True, **kw))
        dens = build_exp_tens(pw, ww, [SIG_P], [1], [False], [False], [0.0], verbose=False)
        ref[i] = entropy_exp_tens(dens, method=method, verbose=False)
    got = windowed_entropy(p_attr, None, [SIG_P, SIG_T], [1, 1], [False, False],
                           [False, False], [0.0, 0.0], centres,
                           context_window=(shape, W_WIDTH), method=method,
                           window_attr=1, drop_window_attr=True, verbose=False)
    assert np.allclose(got, ref, rtol=1e-9, atol=1e-9, equal_nan=True)


def test_entropy_retain_axis(triple):
    p_attr, centres = triple
    ref = np.empty(len(centres))
    for i, c in enumerate(centres):
        pw, ww, _ = unpack_pre_maet(weight_events(p_attr, None, 1, 0, float(c), 1.0,
                                  is_per=False, period=0.0, width=W_WIDTH, drop_input_attr=False))
        dens = build_exp_tens(pw, ww, [SIG_P, SIG_T], [1, 1], [False, False],
                              [False, False], [0.0, 0.0], verbose=False)
        ref[i] = entropy_exp_tens(dens, method="renyi2", verbose=False)
    got = windowed_entropy(p_attr, None, [SIG_P, SIG_T], [1, 1], [False, False],
                           [False, False], [0.0, 0.0], centres,
                           context_window=(1.0, W_WIDTH), method="renyi2",
                           window_attr=1, drop_window_attr=False, verbose=False)
    assert np.allclose(got, ref, rtol=1e-9, atol=1e-9)


def test_entropy_drop_r_ge_2_now_allowed(triple):
    """Dropping a bundled axis is allowed: the centroid `locate` reduces it
    for the window, then it is removed from the density (deletion is no
    longer restricted to r = 1)."""
    p_attr, centres = triple
    got = windowed_entropy(p_attr, None, [SIG_P, SIG_T], [1, 2], [False, False],
                           [False, False], [0.0, 0.0], centres,
                           context_window=(1.0, W_WIDTH), window_attr=1,
                           drop_window_attr=True, verbose=False)
    assert got.shape == (len(centres),)
    assert np.all(np.isfinite(got))


def test_entropy_marginalise_not_implemented(triple):
    """marginalise (integrate out a retained axis) is reserved but unimplemented."""
    p_attr, centres = triple
    with pytest.raises(NotImplementedError):
        windowed_entropy(p_attr, None, [SIG_P, SIG_T], [1, 1], [False, False],
                         [False, False], [0.0, 0.0], centres,
                         context_window=(1.0, W_WIDTH), window_attr=1,
                         drop_window_attr=False, marginalise=0, verbose=False)


def test_entropy_requires_width(triple):
    """No query means no extent to size a default window from, so an explicit
    width is required."""
    p_attr, centres = triple
    with pytest.raises(ValueError):
        windowed_entropy(p_attr, None, [SIG_P, SIG_T], [1, 1], [False, False],
                         [False, False], [0.0, 0.0], centres,
                         context_window=(1.0, None), window_attr=1,
                         drop_window_attr=False, verbose=False)


# --------------------------------------------------------------------------
# windowed_similarity: single-axis surface
# --------------------------------------------------------------------------
@pytest.mark.parametrize("normalize", ["oneSidedDenom", "cosine"])
def test_similarity_locked(triple, query, normalize):
    p_attr, centres = triple
    qv = query[1].ravel(); q_ext = float(qv.max() - qv.min()); mu_q = float(qv.mean())
    ref = np.empty(len(centres))
    for i, c in enumerate(centres):
        pc, wc, _ = unpack_pre_maet(weight_events(p_attr, None, 1, 0, float(c), 1.0,
                                  width=q_ext, is_per=False, period=0.0, drop_input_attr=False))
        pq, wq, _ = unpack_pre_maet(translate_attributes(query, None, [None, np.array([[c - mu_q]])]))
        ref[i] = cos_sim_exp_tens(pc, wc, pq, wq, [SIG_P, SIG_T], [1, 1],
                                  [False, False], [False, False], [0.0, 0.0],
                                  normalize=normalize, verbose=False)
    got = windowed_similarity(p_attr, None, query, None, [SIG_P, SIG_T], [1, 1],
                              [False, False], [False, False], [0.0, 0.0], centres,
                              normalize=normalize, window_attr=1, drop_window_attr=False, verbose=False)
    assert got.shape == (len(centres),)
    assert np.allclose(got, ref, rtol=1e-9, atol=1e-9)


def test_similarity_decoupled_correlogram(triple, query):
    """2-D ``query_centres`` fixes the window at each anchor while the query
    slides across the lags: the anchor x lag correlogram surface."""
    p_attr, centres = triple
    pitch, onset = p_attr
    mu_q = float(query[1].mean())
    anchors = centres[1:8]
    tau = np.linspace(-1.0, 1.0, 9)
    half = 1.5
    ref = np.full((len(anchors), len(tau)), np.nan)
    for ia, a in enumerate(anchors):
        keep = np.abs(onset.ravel() - a) <= half
        pc = [pitch[:, keep], onset[:, keep]]
        for it, t in enumerate(tau):
            pq, wq, _ = unpack_pre_maet(translate_attributes(query, None, [None, np.array([[(a - t) - mu_q]])]))
            ref[ia, it] = cos_sim_exp_tens(pc, None, pq, wq, [SIG_P, SIG_T], [1, 1],
                                           [False, False], [False, False], [0.0, 0.0],
                                           normalize="oneSidedDenom", verbose=False)
    qc2d = anchors[:, None] - tau[None, :]
    got = windowed_similarity(p_attr, None, query, None, [SIG_P, SIG_T], [1, 1],
                              [False, False], [False, False], [0.0, 0.0], anchors,
                              query_centres=qc2d, context_window=(1.0, 2 * half),
                              normalize="oneSidedDenom", window_attr=1, drop_window_attr=False, verbose=False)
    assert got.shape == (len(anchors), len(tau))
    assert np.allclose(got, ref, rtol=1e-9, atol=1e-9, equal_nan=True)


def test_similarity_generative_sweep_matches_explicit(triple, query):
    p_attr, _ = triple
    onset = p_attr[1]
    explicit = np.arange(onset.min(), onset.max() + 1e-9, 1.0)
    g_exp = windowed_similarity(p_attr, None, query, None, [SIG_P, SIG_T], [1, 1],
                                [False, False], [False, False], [0.0, 0.0], explicit,
                                normalize="oneSidedDenom", window_attr=1, drop_window_attr=False, verbose=False)
    g_gen = windowed_similarity(p_attr, None, query, None, [SIG_P, SIG_T], [1, 1],
                                [False, False], [False, False], [0.0, 0.0],
                                start=float(onset.min()), stop=float(onset.max()), step=1.0,
                                normalize="oneSidedDenom", window_attr=1, drop_window_attr=False, verbose=False)
    assert g_gen.shape == g_exp.shape
    assert np.allclose(g_gen, g_exp, rtol=1e-9, atol=1e-9)


def test_centres_and_generative_mutually_exclusive(triple, query):
    p_attr, centres = triple
    with pytest.raises(ValueError):
        windowed_similarity(p_attr, None, query, None, [SIG_P, SIG_T], [1, 1],
                            [False, False], [False, False], [0.0, 0.0], centres,
                            step=1.0, window_attr=1, drop_window_attr=False, verbose=False)


# --------------------------------------------------------------------------
# windowed_similarity: multi-axis surface and locate
# --------------------------------------------------------------------------
def test_single_axis_equals_one_entry_sweep(triple, query):
    """The single-axis surface is exactly the one-entry multi-axis form."""
    p_attr, centres = triple
    single = windowed_similarity(p_attr, None, query, None, [SIG_P, SIG_T], [1, 1],
                                 [False, False], [False, False], [0.0, 0.0], centres,
                                 normalize="oneSidedDenom", window_attr=1,
                                 drop_window_attr=False, verbose=False)
    multi = windowed_similarity(p_attr, None, query, None, [SIG_P, SIG_T], [1, 1],
                                [False, False], [False, False], [0.0, 0.0],
                                sweep={1: centres}, drop={1: False},
                                normalize="oneSidedDenom", verbose=False)
    assert np.allclose(single, multi, rtol=1e-12, atol=1e-12)


def test_multi_axis_two_dim_map(triple, query):
    """Sweeping pitch (compared) and time (dropped) yields a 2-D map."""
    p_attr, centres = triple
    pitch = p_attr[0]
    p_centroid = float(query[0].mean())
    pgrid = p_centroid + np.arange(-4.0, 5.0, 2.0)
    R = windowed_similarity(p_attr, None, query, None, [SIG_P, SIG_T], [1, 1],
                            [False, False], [False, False], [0.0, 0.0],
                            sweep={1: centres, 0: pgrid}, drop={1: True, 0: False},
                            context_window={0: {"shape": "rect", "width": 6.0}},
                            normalize="oneSidedDenom", verbose=False)
    assert R.shape == (len(centres), len(pgrid))
    assert np.all(np.isfinite(R))


def test_locate_is_wired():
    """On an asymmetric bundle, 'centroid' and 'start' place the window (and
    the query translation) differently, so the self-similarity sweep peaks at
    different centres and the two arrays must not coincide."""
    qp = np.array([[60., 64.]]); qo = np.array([[0.0, 0.6]])   # centroid 0.3, start 0.0
    qb, qw, qs = unpack_pre_maet(bind_events([qp, qo], None, [1, 2], step=1, rel_outer=[False, False]))
    centres = np.linspace(-0.4, 0.7, 12)        # keeps both window centres non-empty
    common = dict(window_attr=1, drop_window_attr=False, context_window=(1.0, 1.5),
                  normalize="cosine", specs=qs, verbose=False)
    a = windowed_similarity(qb, qw, qb, qw, [SIG_P, SIG_T], [1, 2], [False, False],
                            [False, False], [0.0, 0.0], centres, locate="centroid", **common)
    b = windowed_similarity(qb, qw, qb, qw, [SIG_P, SIG_T], [1, 2], [False, False],
                            [False, False], [0.0, 0.0], centres, locate="start", **common)
    assert a.shape == b.shape == (len(centres),)
    # both attain a clear self-similarity peak, but at different centres
    assert np.nanmax(a) > 0.9 and np.nanmax(b) > 0.9
    assert abs(centres[np.nanargmax(a)] - centres[np.nanargmax(b)]) > 0.2
    assert not np.allclose(a, b, equal_nan=True)


def _qw_setup():
    """Asymmetric query: two close events plus one far one, so a narrow query
    window genuinely reshapes it rather than rescaling it uniformly."""
    pc = [np.array([[0., 100., 200., 300., 400., 500.]]),
          np.array([[0., 1., 2., 3., 4., 5.]])]
    wc = [np.ones((1, 6)), np.ones((1, 6))]
    pq = [np.array([[0., 100., 500.]]), np.array([[0., 1., 4.]])]
    wq = [np.ones((1, 3)), np.ones((1, 3))]
    common = dict(centres=np.arange(6.0), window_attr=1, drop_window_attr=False,
                  normalize="cosine", verbose=False)
    run = lambda **kw: np.asarray(windowed_similarity(
        pc, wc, pq, wq, [30.0, 0.25], [1, 1], [False, False], [False, False],
        [0.0, 0.0], **common, **kw))
    return run


def test_query_window_is_applied():
    """A query window must reshape the query. Regression: the argument was
    accepted, validated, and threaded to the worker, but never read, so a
    caller asking for a windowed query silently got an unwindowed one."""
    run = _qw_setup()
    plain = run(context_window=("gauss", 2.0))
    narrow = run(context_window=("gauss", 2.0), query_window=("gauss", 0.8))
    assert not np.allclose(plain, narrow)
    # Suppressing the far query event leaves the near pair, which matches the
    # context far better at its best offset.
    assert np.nanmax(narrow) > np.nanmax(plain)
    assert narrow.min() >= 0.0 and narrow.max() <= 1.0 + 1e-12


def test_wide_rect_query_window_is_an_exact_no_op():
    """A rectangle wider than the query's own extent includes every event at
    unit weight, so it must reproduce the unwindowed profile exactly."""
    run = _qw_setup()
    plain = run(context_window=("gauss", 2.0))
    wide = run(context_window=("gauss", 2.0), query_window=("rect", 1000.0))
    assert np.array_equal(plain, wide)


def test_gaussian_query_window_converges_to_no_op():
    """Widening a Gaussian query window must approach the unwindowed profile,
    and at the O(1/sd^2) rate set by the window's curvature across the query."""
    run = _qw_setup()
    plain = run(context_window=("gauss", 2.0))
    d = [float(np.max(np.abs(run(context_window=("gauss", 2.0),
                                 query_window=("gauss", sd)) - plain)))
         for sd in (100.0, 1000.0, 10000.0)]
    assert d[0] > d[1] > d[2]
    for lo, hi in zip(d[1:], d[:-1]):
        assert 50.0 < hi / lo < 200.0
